agent/consul/leader_connect_ca_test.go

```diff
@@ -0,0 +1,256 @@
+package consul
+
+import (
+	"context"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/consul/agent/connect"
+	ca "github.com/hashicorp/consul/agent/connect/ca"
+	"github.com/hashicorp/consul/agent/consul/state"
+	"github.com/hashicorp/consul/agent/metadata"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/sdk/testutil"
+	"github.com/hashicorp/go-version"
+	"github.com/hashicorp/serf/serf"
+	"github.com/stretchr/testify/require"
+)
+
+type mockCAServerDelegate struct {
+	t           *testing.T
+	config      *Config
+	store       *state.Store
+	primaryRoot *structs.CARoot
+	callbackCh  chan string
+}
+
+func NewMockCAServerDelegate(t *testing.T, config *Config) *mockCAServerDelegate {
+	delegate := &mockCAServerDelegate{
+		t:           t,
+		config:      config,
+		store:       state.NewStateStore(nil),
+		primaryRoot: connect.TestCAWithTTL(t, nil, 1*time.Second),
+		callbackCh:  make(chan string, 0),
+	}
+	delegate.store.CASetConfig(1, testCAConfig())
+
+	return delegate
+}
+
+func (m *mockCAServerDelegate) State() *state.Store {
+	return m.store
+}
+
+func (m *mockCAServerDelegate) IsLeader() bool {
+	return true
+}
+
+func (m *mockCAServerDelegate) CheckServers(datacenter string, fn func(*metadata.Server) bool) {
+	ver, _ := version.NewVersion("1.6.0")
+	fn(&metadata.Server{
+		Status: serf.StatusAlive,
+		Build:  *ver,
+	})
+}
+
+func (m *mockCAServerDelegate) ApplyCARequest(req *structs.CARequest) (interface{}, error) {
+	return ca.ApplyCARequestToStore(m.store, req)
+}
+
+func (m *mockCAServerDelegate) createCAProvider(conf *structs.CAConfiguration) (ca.Provider, error) {
+	return &mockCAProvider{
+		callbackCh: m.callbackCh,
+		rootPEM:    m.primaryRoot.RootCert,
+	}, nil
+}
+
+func (m *mockCAServerDelegate) forwardDC(method, dc string, args interface{}, reply interface{}) error {
+	switch method {
+	case "ConnectCA.Roots":
+		roots := reply.(*structs.IndexedCARoots)
+		roots.TrustDomain = connect.TestClusterID
+		roots.Roots = []*structs.CARoot{m.primaryRoot}
+		roots.ActiveRootID = m.primaryRoot.ID
+	case "ConnectCA.SignIntermediate":
+		r := reply.(*string)
+		*r = m.primaryRoot.RootCert
+	}
+
+	m.callbackCh <- fmt.Sprintf("forwardDC/%s", method)
+
+	return nil
+}
+
+func (m *mockCAServerDelegate) generateCASignRequest(csr string) *structs.CASignRequest {
+	return &structs.CASignRequest{
+		Datacenter: m.config.PrimaryDatacenter,
+		CSR:        csr,
+	}
+}
+
+func (m *mockCAServerDelegate) raftApply(t structs.MessageType, msg interface{}) (interface{}, error) {
+	if t == structs.ConnectCARequestType {
+		req := msg.(*structs.CARequest)
+		act, err := m.store.CARootSetCAS(1, req.Index, req.Roots)
+		require.NoError(m.t, err)
+		require.True(m.t, act)
+
+		act, err = m.store.CACheckAndSetConfig(1, req.Config.ModifyIndex, req.Config)
+		require.NoError(m.t, err)
+		require.True(m.t, act)
+	}
```

Comment:
Might want an `else` here that just returns an obvious error to avoid future test headaches.